Use batching in `BackgroundDataStoreProcessor` processes

cyperdark opened the pull request:
- Added `BulkWrite` method to realm
- Added `TryLookup` method to reusable sqlite connection in `LocalCachedBeatmapMetadataSource`

<br>

### benchmark (updated 2026-01-20)

|   | star ratings (45492) | object counts (45492) | scores statistics (3928) | total score (3928) | ranks (3928) | submission and rank dates (9695) | user tags (25412) |
| ------------- | ------------- | ------------- | ------------- | ------------- | ------------- | ------------- | ------------- |
| Before | 4111953ms | 1766952ms | 103199ms | 104935ms | 101620ms | 279318ms | 770134ms |
| After | 2286840ms | 391678ms | 2221ms | 4765ms | 257ms | 5918ms | 10191ms |
| Speed up | 1.798x | 4.511x | 46.465x | 22.022x | 395.408x | 47.198x | 75.570x |


Tests done offline on HDD WD20EZBX - 275Mb/s(?)

<br>

## Notes

- On these changes i was getting game freezes for a 20-40s due to `backpopulateMissingSubmissionAndRankDates` method sending 9k sets to `RealmDetachedBeatmapStore` subscription
- ~~TryLookup in sqlite have duplicated code, since i wasnt sure if i allowed to touch this one~~
- I didnt touch `processOnlineBeatmapSetsWithNoUpdate` because `beatmapUpdater.Process` used in other place and im not if i should touch it.
- i tried to make benchmark project but couldnt because i lack knowledge of codebase and c#
- https://github.com/ppy/osu/pull/36128/changes/dd889ea5107a4a716ba4ebeb1fddd6cd24ecda18, might be bad idea but not sure

cyperdark committed:
Add measuring to `BackgroundDataStoreProcessor`

peppy commented:
Did you try performing any operations when testing this? It should deadlock. Realm doesn't support multiple concurrent writes, so holding writes open for long periods as you are doing here, would under my presumptions, cause something like running gameplay (and triggering an update-thread-bound write) to freeze the game.

cyperdark commented:
when it was recalculating beatmaps i was scrolling though song select and it was working fine, let me test it gameplay and song select once again. If it wont work i'll try another approach i have in mind

cyperdark commented:
Okay i've tested and it's not adding maps to collection, doesnt switch from gameplay to result screen and i assume it wont allow/will freeze if i try to download new map

Another approach is writing once at the end of loop or let's say every 500 beatmaps, like this:
<img width="1847" height="967" alt="image" src="https://github.com/user-attachments/assets/c59764bb-5acf-4226-aae0-9048e447c9ed" />

cyperdark commented:
hmm, best candidates for testing that people who have their lazer files/ and realm.client on HDD.

So using bulk update by 100 beatmaps in `populateMissingStarRatings` is questionable fix but what about `backpopulateUserTags`, it's less resource demanding task where this approach will be better than updating each beatmap, especially when people like me have over 10k beatmaps and it takes 6 minutes to update all of them

cyperdark commented:
idk how to measure overhead, but here's video with performance graph with calculating star ratings and scrolling in song select. Closer to the end i rerun calculations for small chunk of maps and then started updating tags (both used bulk, star ratings 100, user tags 500): https://youtu.be/m_7wohmkI5U


with only user tags updating: https://youtu.be/6X7xrV1VsyA

edit: but to keep in mind that this tests made on high end pc

cihe13375 commented:
> I want to understand why. Not hack around it.

I don't really know how realm works internally, but the realm .net documentation (https://github.com/realm/realm-dotnet/blob/community/Guides/crud/write-transactions.md?plain=1) says such a thing:
> When Realm commits a transaction, Realm writes all changes to disk.

As each map processed calls `realmAccess.Write` (which then calls a transaction commit), the disk write is apparently triggered at every map.

I tried silly a local change that removed other operations in the user tag population:
```diff
diff --git a/osu.Game/Database/BackgroundDataStoreProcessor.cs b/osu.Game/Database/BackgroundDataStoreProcessor.cs
index f8d1b9ae51..8665ce3dab 100644
--- a/osu.Game/Database/BackgroundDataStoreProcessor.cs
+++ b/osu.Game/Database/BackgroundDataStoreProcessor.cs
@@ -655,8 +655,6 @@ private void backpopulateUserTags()
 
             if (metadataSourceFetchDate <= lastPopulation)
             {
-                Logger.Log($@"Skipping user tag population because the local metadata source hasn't been updated since the last time user tags were checked ({lastPopulation.Value:d})");
-                return;
             }
 
             Logger.Log(@"Querying for beatmaps that do not have user tags");
@@ -695,27 +693,6 @@ private void backpopulateUserTags()
                     // ReSharper disable once MethodHasAsyncOverload
                     realmAccess.Write(r =>
                     {
-                        BeatmapInfo beatmap = r.Find<BeatmapInfo>(id)!;
-
-                        bool lookupSucceeded = localMetadataSource.TryLookup(beatmap, out var result);
-
-                        if (lookupSucceeded)
-                        {
-                            Debug.Assert(result != null);
-
-                            var userTags = result.UserTags.ToHashSet();
-
-                            if (!userTags.SetEquals(beatmap.Metadata.UserTags))
-                            {
-                                beatmap.Metadata.UserTags.Clear();
-                                beatmap.Metadata.UserTags.AddRange(userTags);
-                                return true;
-                            }
-
-                            return false;
-                        }
-
-                        Logger.Log(@$"Could not find {beatmap.GetDisplayString()} in local cache while backpopulating missing user tags");
                         return false;
                     });
 
```
The performance is just as bad, and I can clearly hear the sound of HDD working during the stub "user tag population" process.

If I move client.realm to SSD and create a symlink at the osu data folder on HDD, all problems are gone.

cyperdark commented:
> I want to understand why. Not hack around it.

because of high disk usage, here's video with disk usage graphs https://youtu.be/j5MrpL_q0c0
p.s. new folder 10 is regular lazer version, new folder 11 is pr version

as you can see in video a regular lazer version does too much disk writes especially on "User Tags Population" process. I also think that for user tags process i can increase bulk size to 2000-3000 since this process is way faster and it'll use less disk which will both good for lazer and disk itself

cyperdark commented:
locally i reverted PR and added notifications to `RealmDetachedBeatmapStore.Update` and here you can see (probably) why people had lag spikes: https://youtu.be/FZQfdlUM0Ic

cyperdark commented:
> Also sharing builds like this on reddit is a HUGE security no-no, I'd suggest removing your post IMMEDIATELY.

deleted

> If you spend another 30 minutes blaming this class

not on this class, it's in general to find a place which causes things to lag spike

> The issue is that for some users, realm operations are 100-1000x slower than the average osu! user. For no good reason.

oh i see, but in this PR i am trying to speed processes by using bulk, because one day i was curious what exactly `Populate User Tags` process doing, i looked at network requests and there was none of them to osu api, then i checked source code and saw that it's writing changes on every iteration and since i used mongodb where you have function BulkWrite i thought i can add something similar to lazer. 

p.s. i have uncommitted changes which properly adds BulkWrite

So can this PR still be merged with Bulk writes idea or it's not needed?

peppy commented:
> So can this PR still be merged with Bulk writes idea or it's not needed?

It needs validation testing by myself or someone else on the team with a sound approach to testing. This may take weeks to months, so please chill.

If your `BulkWrite` change reduces the LoC changed then it may be a win. Right now this is verging on "very hard to review" size of change.

cyperdark committed:
Add `BulkWrite` method to Realm

peppy commented on osu.Game/Beatmaps/LocalCachedBeatmapMetadataSource.cs:
```diff
@@ -146,6 +146,68 @@ public bool TryLookup(BeatmapInfo beatmapInfo, [NotNullWhen(true)] out OnlineBea
             }
         }
 
+        public bool TryLookup(SqliteConnection db, int version, BeatmapInfo beatmapInfo, [NotNullWhen(true)] out OnlineBeatmapMetadata? onlineMetadata)
```
What on earth? Is this a copy paste of the method above? Why?

peppy reviewed (changes requested):
not even reviewing most of the code, just a few things

cyperdark commented on osu.Game/Beatmaps/LocalCachedBeatmapMetadataSource.cs:
```diff
@@ -162,7 +224,7 @@ private void tryPurgeCache()
         }
 
-        private SqliteConnection getConnection() =>
+        public SqliteConnection GetConnection() =>
```
i manually did testing:
1. shutdown pc => start pc
2. run lazer before sqlite changes
3. shutdown pc => start pc
4. run lazer with sqlite changes

timings copied for runtime.log which i added in [this commit](https://github.com/ppy/osu/pull/36128/changes/82ac081588722461b5b476d3ea30812a29001ba5)

cyperdark replied:
oh shit i think i used from lazer versions, because i did more test and difference on HDD is significant compare to one above, and this time i shutdown pc for a 5 sec and then run other version:

SSD MZ-V8P1T0BW - 7000Mb/5000Mb

| amount | 9057 | 43498 |
| ------------- | ------------- | ------------- |
| before | 3475ms | 8371ms |
| after | 2915ms | 6475ms |
| speed up | 1.192x | 1.292x |

<br>

HDD WD20EZBX - 275Mb/s(?)

| amount | 9057 | 43437 |
| ------------- | ------------- | ------------- |
| before | 40388ms | 49045ms |
| after | 15256ms | 23023ms |
| speed up | 2.647x | 2.130x |

<br>

log files: 

[1768928271.runtime.log](https://github.com/user-attachments/files/24743214/1768928271.runtime.log)
[1768928461.runtime.log](https://github.com/user-attachments/files/24743215/1768928461.runtime.log)

cyperdark replied:
i think i misunderstood this conversation, i thought i need to make benchmark to test all methods in `BackgroundDataStoreProcessor`, and reading conversation again i think it meant to benchmark sqlite connection difference. Here results:

| Method                  | Amount | Mean       | Error    | StdDev   | Ratio | RatioSD | Gen0       | Gen1       | Allocated | Alloc Ratio |
|------------------------ |------- |-----------:|---------:|---------:|------:|--------:|-----------:|-----------:|----------:|------------:|
| BenchmarkWithoutReusing | 1000   |   288.7 ms |  5.68 ms |  5.31 ms |  1.00 |    0.03 |  6000.0000 |  5000.0000 | 105.76 MB |        1.00 |
| BenchmarkWithReusing    | 1000   |   184.0 ms |  3.01 ms |  2.67 ms |  0.64 |    0.01 |  5500.0000 |  4500.0000 | 103.52 MB |        0.98 |
|                         |        |            |          |          |       |         |            |            |           |             |
| BenchmarkWithoutReusing | 10000  | 2,028.7 ms | 14.52 ms | 12.87 ms |  1.00 |    0.01 | 27000.0000 | 12000.0000 | 440.16 MB |        1.00 |
| BenchmarkWithReusing    | 10000  | 1,031.4 ms |  7.65 ms |  7.16 ms |  0.51 |    0.00 | 25000.0000 | 11000.0000 | 418.01 MB |        0.95 |

<details>
  <summary>benchmark code</summary>

```c#
// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
// See the LICENCE file in the repository root for full licence text.

using BenchmarkDotNet.Attributes;
using osu.Game.Beatmaps;
using osu.Framework.Testing;
using System;
using osu.Game.Database;
using osu.Game.Tests.Resources;
using System.Collections.Generic;
using System.Linq;
using osu.Framework.Utils;

namespace osu.Game.Benchmarks
{
    public class BenchmarkLocalCachedMetadataSource : BenchmarkTest
    {
        private TemporaryNativeStorage storage = null!;
        private LocalCachedBeatmapMetadataSource localMetadataSource = null!;
        private RealmAccess realm = null!;

        [Params(1_000, 10_000)]
        public int Amount { get; set; }

        public override void SetUp()
        {
            storage = new TemporaryNativeStorage($"sqlite");
            realm = new RealmAccess(storage, OsuGameBase.CLIENT_DATABASE_FILENAME);

            int totalCount = realm.Run(r => r.All<BeatmapInfo>().Count());
            while (totalCount < Amount)
            {
                var baseTestBeatmap = CreateTestBeatmapSetInfo(RNG.Next(1, 1000), true);
                realm.Write(r => r.Add(baseTestBeatmap, true));

                totalCount = realm.Run(r => r.All<BeatmapInfo>().Count());
                Console.WriteLine($"Generating maps: {totalCount}/{Amount}");
            }

            localMetadataSource = new LocalCachedBeatmapMetadataSource(storage);
        }


        [Benchmark(Baseline = true)]
        public void BenchmarkWithoutReusing()
        {
            int found = 0;
            int notFound = 0;

            var beatmapIds = realm.Run(r => new HashSet<Guid>(r.All<BeatmapInfo>().AsEnumerable().Select(s => s.ID))).Take(Amount).ToList();

            realm.Run(r =>
            {
                foreach (Guid id in beatmapIds)
                {
                    var beatmap = r.Find<BeatmapInfo>(id);
                    if (beatmap == null) continue;

                    bool lookupSucceeded = localMetadataSource.TryLookup(null, null, beatmap, out var result);
                    if (lookupSucceeded && beatmap.OnlineID == result?.BeatmapID)
                        found++;
                    else
                        notFound++;
                }
            });
        }

        [Benchmark]
        public void BenchmarkWithReusing()
        {
            var sqliteConnection = localMetadataSource.GetConnection();
            sqliteConnection.Open();

            int sqliteVersion = localMetadataSource.GetCacheVersion(sqliteConnection);

            int found = 0;
            int notFound = 0;

            var beatmapIds = realm.Run(r => new HashSet<Guid>(r.All<BeatmapInfo>().AsEnumerable().Select(s => s.ID))).Take(Amount).ToList();

            realm.Run(r =>
            {
                foreach (Guid id in beatmapIds)
                {
                    var beatmap = r.Find<BeatmapInfo>(id);
                    if (beatmap == null) continue;

                    bool lookupSucceeded = localMetadataSource.TryLookup(sqliteConnection, sqliteVersion, beatmap, out var result);
                    if (lookupSucceeded && beatmap.OnlineID == result?.BeatmapID)
                        found++;
                    else
                        notFound++;
                }
            });
        }

        // yoinked from TestPerformanceWithManyBeatmaps
        protected static BeatmapSetInfo CreateTestBeatmapSetInfo(int? fixedDifficultiesPerSet, bool randomMetadata)
        {
            var beatmapSetInfo = TestResources.CreateTestBeatmapSetInfo(fixedDifficultiesPerSet ?? RNG.Next(1, 4));

            if (randomMetadata)
            {
                char randomCharacter = getRandomCharacter();

                var metadata = new BeatmapMetadata
                {
                    Artist = $"{randomCharacter}ome Artist " + RNG.Next(0, 9),
                    Title = $"{randomCharacter}ome Song (set id {beatmapSetInfo.OnlineID:000}) {Guid.NewGuid()}",
                    Author = { Username = $"{randomCharacter}ome Guy " + RNG.Next(0, 9) },
                };

                foreach (var beatmap in beatmapSetInfo.Beatmaps)
                    beatmap.Metadata = metadata.DeepClone();
            }

            return beatmapSetInfo;
        }

        private static long randomCharPointer;
        private static char getRandomCharacter()
        {
            const string chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz*";
            return chars[(int)((randomCharPointer++ / 2) % chars.Length)];
        }
    }
}
```
</details>

cyperdark committed:
Add reusable sqlite connection for `TryLookup`

cyperdark committed:
Use batching in `BackgroundDataStoreProcessor`

cyperdark committed:
Add batching to `processBeatmapsWithMissingObjectCounts`

cyperdark committed:
Add batching to `upgradeModMultipliers`

cyperdark committed:
Move actions outside of methods

cyperdark committed:
Make chunk_size a method argument

cyperdark committed:
Simplify background processors by using single `realmAccess.Run` call

cyperdark committed:
Minor fixes to logs messages

cyperdark committed:
Fix inspectcode issues

cyperdark committed:
Cleanup

cyperdark committed:
Refactor `BackgroundDataStoreProcessor` to use helper method

cyperdark committed:
Remove repeated code

peppy commented on osu.Game/Beatmaps/BeatmapUpdater.cs:
```diff
@@ -71,20 +71,17 @@ public void Process(BeatmapSetInfo beatmapSet, MetadataLookupScope lookupScope =
 
         public void ProcessObjectCounts(BeatmapInfo beatmapInfo, MetadataLookupScope lookupScope = MetadataLookupScope.LocalCacheFirst)
         {
-            beatmapInfo.Realm!.Write(_ =>
```
?

this is documented nowhere and now does not match the other methods exposed here. no reasoning, nothing.

cyperdark replied:
oh i just realized i fucked up here, I was rewriting `processBeatmapsWithMissingObjectCounts` to use batch method and initially moved entire code from `ProcessObjectCounts` to `processBeatmapsWithMissingObjectCounts` but then for some reason i decided to remove `beatmapInfo.Realm!.Write` from `ProcessObjectCounts` and i use it instead of copying code. But now after reading comments in that method i see why what i did is bad.

right now im thinking about how i can fix `processBeatmapsWithMissingObjectCounts`, and i though about two ways:
1. move calculation of object count to `processBeatmapsWithMissingObjectCounts` and remove `ProcessObjectCounts` from `BeatmapUpdater`
2. remove batch processing from `processBeatmapsWithMissingObjectCounts`

since i dont know much about how `workingBeatmapCache` works, i assume if i call 500 times `workingBeatmapCache.Invalidate` it would cause significant lag but not sure

peppy commented on osu.Game/Database/BackgroundDataStoreProcessor.cs:
```diff
@@ -153,118 +154,101 @@ private void clearOutdatedStarRatings()
         /// Star rating recalculations can be ran strictly locally.
         /// </remarks>
-        private void populateMissingStarRatings()
+        private void populateMissingStarRatings(int chunkSize = 500)
```
Remove argument, it's never used.

cyperdark replied:
oh shit i used it for local tests and forgot to remove

peppy commented on osu.Game/Database/BackgroundDataStoreProcessor.cs:
```diff
@@ -280,331 +264,226 @@ private void processOnlineBeatmapSetsWithNoUpdate()
+                        RunningNotificationTemplate = "Upgrading scores to new mod multipliers",
+                        CompletedNotificationTemplate = "scores have been upgraded to the new mod multipliers",
+                        ExceptionTemplate = "Failed to upgrade mod multipliers",
+                        FinishedTemplate = "Upgrading {processed} of {total} scores to new mod multipliers completed in {elapsed}ms"
```
what is this templating?

cyperdark replied:
i have troubles with naming things so i left this as temporary name, it's for logger/notification messages in `batchedProcessing`, since each method had different messages text for logs and notification and i couldn't just use one message for everything i decided to put them in function argument (found online that you can use record for this in c#, dont know if it's right one here)

peppy reviewed (commented):
is this ai? it's going more and more off the rails
